proxystore/store/dim/ucx.py

```
@@ -0,0 +1,305 @@
+            server_process.start()
+
+            # allocate some time to start the server process
+            sleep(0.2)
+
+        super().__init__(
+            name,
+            cache_size=cache_size,
+            stats=stats,
+            kwargs={'interface': interface, 'port': self.port},
+        )
+
+    def _start_server(self) -> None:
+        """Launch the local UCX server (Peer) process."""
+        logger.info(
+            f'starting server on host {self.host} with port {self.port}',
+        )
+
+        # create server
+        ps = UCXServer(self.host, self.port)
+        asyncio.run(ps.launch())
+
+        logger.info(f'Server running at address {self.addr}')
+
+    def create_key(self, obj: Any) -> UCXStoreKey:
+        return UCXStoreKey(
+            ucx_key=utils.create_key(obj),
+            obj_size=len(obj),
+            peer=self.addr,
+        )
+
+    def evict(self, key: UCXStoreKey) -> None:
+        logger.debug(f'Client issuing an evict request on key {key}.')
+
+        event = serialize({'key': key.ucx_key, 'data': None, 'op': 'evict'})
+        self._loop.run_until_complete(self.handler(event, key.peer))
+
+        self._cache.evict(key)
+
+    def exists(self, key: UCXStoreKey) -> bool:
+        logger.debug(f'Client issuing an exists request on key {key}.')
+
+        event = serialize(
+            {'key': key.ucx_key, 'data': None, 'op': 'exists'},
+        )
+        return deserialize(
+            self._loop.run_until_complete(self.handler(event, key.peer)),
+        )
+
+    def get_bytes(self, key: UCXStoreKey) -> bytes | None:
+        res: bytes | None
+        logger.debug(f'Client issuing get request on key {key}.')
+
+        event = serialize({'key': key.ucx_key, 'data': '', 'op': 'get'})
+        res = self._loop.run_until_complete(self.handler(event, key.peer))
+
+        try:
+            s = deserialize(res)
+
+            if isinstance(s, Status) and not s.success:
```

`Status` is checked here, do the other operations also need to check `Status`?

---

Yes, but I don't think the other stores throw errors just yet. I'd need to add test cases in `store_*` if the dim need to raise errors, unless we ignore KeyErrors, but raise for every other error? I still need to determine the scope of the errors that can arise with ucx, margo and websockets

---

Yeah, KeyErrors in this case are always safe and expected. For UXCStore it seems okay as is since the UCXServer will crash on exceptions. Websockets looks the same as UCX, but MargoServer catches all exceptions so those need to be handled somewhere.

Since we don't know the scope of errors yes we should default to raising everything. As we understand them better we can more narrowly handle/suppress errors as needed.